uvicorn/_async_agnostic/server.py

```diff
@@ -0,0 +1,268 @@
+                    async with (
+                        self._backend.start_soon(self._main_loop),
+                        self._backend.start_soon(self._listen_signals),
+                    ):
+                        # Server has started.
+                        self._sync_started_event.set()
+                        self._log_started_message(listeners, sockets=sockets)
+                        # Let the server run until exit is requested.
+                        await shutdown_trigger()
+                        await self._shutdown_event.set()
+            finally:
+                if not self._force_exit:
+                    await lifespan.shutdown()
+
+        message = "Finished server process [%d]"
+        color_message = "Finished server process [" + click.style("%d", fg="cyan") + "]"
+        self._logger.info(
+            "Finished server process [%d]",
+            process_id,
+            extra={"color_message": color_message},
+        )
+
+    async def _serve(
+        self,
+        sockets: List[socket.SocketType] = None,
+        *,
+        task_status: TaskStatus = TaskStatus.IGNORED,
+    ) -> None:
+        await self._backend.serve_tcp(
+            handle_http11,
+            self._state,
+            self._config,
+            sockets=sockets,
+            wait_close=self._shutdown_event.wait,
+            on_close=self._on_close,
+            task_status=task_status,
+        )
+
+    async def _main_loop(self) -> None:
+        counter = 0
+        should_exit = await self._tick(counter)
+        while not should_exit:
+            counter += 1
+            counter = counter % 864000
+            await self._backend.sleep(0.1)
+            should_exit = await self._tick(counter)
+        await self._shutdown_event.set()
+
+    async def _tick(self, counter: int) -> bool:
+        state = self._state
+        config = self._config
+
+        # Update the default headers, once per second.
+        if counter % 10 == 0:
+            current_time = time.time()
+            current_date = to_internet_date(current_time).encode()
+            state.default_headers = [(b"date", current_date)] + config.encoded_headers
+
+            # Callback to `callback_notify` once every `timeout_notify` seconds.
+            if config.callback_notify is not None:
+                if current_time - self._last_notified > config.timeout_notify:
+                    self._last_notified = current_time
+                    await config.callback_notify()
+
+        # Determine if we should exit.
+        if self._shutdown_event.is_set():
+            return True
+        if config.limit_max_requests is not None:
+            return state.total_requests >= config.limit_max_requests
+        return False
+
+    async def _listen_signals(self) -> None:
+        if threading.current_thread() is not threading.main_thread():
+            await self._closed_event.wait()
+            return
```

Comment:
The existing `Server` doesn't take this into account and blindly tries to listen on signals in all cases, which means any people that want to run Uvicorn in a thread need to deal with this problem. Hence overriding `.install_signal_handlers()` to a "do-nothing" implementation in a tests setup. We could consider an enhancement to the existing `Server` to do a check like this by default in `.install_signal_handlers()`.